atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolTest.java

```diff
@@ -467,6 +470,82 @@ public void throwsWhenTryingToAddServerPresentInCurrentServers() {
                         SafeArg.of("currentServers", CassandraLogHelper.collectionOfHosts(poolServers.keySet())));
     }
 
+    @Test
+    public void tryGettingNewHostsWithDifferentTopologyInvalidatesOnlyNewContainers() {
+        setupThriftServers(Set.of(CASS_SERVER_1.proxy(), CASS_SERVER_2.proxy()));
+
+        CassandraClientPoolingContainer container1 = mock(CassandraClientPoolingContainer.class);
+        CassandraClientPoolingContainer container2 = mock(CassandraClientPoolingContainer.class);
+
+        Map<CassandraServer, CassandraServerOrigin> serversToAdd =
+                Map.of(CASS_SERVER_1, CassandraServerOrigin.TOKEN_RANGE);
+        Map<CassandraServer, CassandraClientPoolingContainer> allContainers = Map.of(
+                CASS_SERVER_1, container1,
+                CASS_SERVER_2, container2);
+
+        CassandraClientPoolImpl pool = createClientPool();
+
+        when(cassandraTopologyValidator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        eq(serversToAdd), eq(allContainers), any(), any()))
+                .thenThrow(new RuntimeException());
+
+        assertThatCode(() -> pool.tryGettingNewHostsWithDifferentTopologyOrInvalidateNewContainersAndThrow(
+                        serversToAdd, allContainers))
+                .isInstanceOf(RuntimeException.class);
+        verify(container1).shutdownPooling();
+        verify(container2, never()).shutdownPooling();
+    }
+
+    @Test
+    public void tryGettingNewHostsWithDifferentTopologyInvalidatesSecondContainerEvenIfFirstOneIsFailed() {
+        setupThriftServers(Set.of(CASS_SERVER_1.proxy(), CASS_SERVER_2.proxy()));
+
+        CassandraClientPoolingContainer container1 = mock(CassandraClientPoolingContainer.class);
+        CassandraClientPoolingContainer container2 = mock(CassandraClientPoolingContainer.class);
+
+        Map<CassandraServer, CassandraServerOrigin> serversToAdd =
+                Map.of(CASS_SERVER_1, CassandraServerOrigin.TOKEN_RANGE, CASS_SERVER_2, CassandraServerOrigin.CONFIG);
+        Map<CassandraServer, CassandraClientPoolingContainer> allContainers =
+                Map.of(CASS_SERVER_1, container1, CASS_SERVER_2, container2);
+
+        CassandraClientPoolImpl pool = createClientPool();
+
+        when(cassandraTopologyValidator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        eq(serversToAdd), eq(allContainers), any(), any()))
+                .thenThrow(new RuntimeException());
+        doThrow(new RuntimeException()).when(container1).shutdownPooling();
```

Comment:
good thinking! that said, I think we might want to also throw this from `container2`'s shutdow - in a `Map.of(...)` the iteration order is not guaranteed and subject to change